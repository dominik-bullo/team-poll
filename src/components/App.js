import { useEffect } from "react";
import { connect } from "react-redux";
import { handleInitialData } from "../actions/shared";
import { Route, Routes } from "react-router-dom";
import LoadingBar from "react-redux-loading-bar";
import Nav from "./Nav";
import Dashboard from "./Dashboard";

function App(props) {
  useEffect(() => {
    props.dispatch(handleInitialData());
  }, []);

  return (
    <>
      <Nav />
      <LoadingBar />
      {props.loading === true ? null : (
        <Routes>
          <Route path="/" exact element={<Dashboard />} />
        </Routes>
      )}
    </>
  );
}

const mapStateToProps = ({ authedUser }) => ({
  loading: authedUser === null,
});

export default connect(mapStateToProps)(App);
