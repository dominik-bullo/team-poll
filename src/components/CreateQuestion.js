import { useState } from "react";
import { connect } from "react-redux";
import { handleAddQuestion } from "../actions/shared";
import { useNavigate } from "react-router-dom";

const CreateQuestion = ({ id, dispatch }) => {
  const [options, setOptions] = useState(["", ""]);
  const navigate = useNavigate();

  const handleChange = (index) => (e) => {
    const newOptions = options.map((option, i) => {
      if (i === index) {
        return e.target.value;
      }
      return option;
    });
    setOptions(newOptions);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleAddQuestion(options[0], options[1], id));
    navigate("/");
  };

  return (
    <div className="pure-u-1 new-question center">
      <h2>New Question</h2>
      <h3>Would you rather...</h3>

      <form className="pure-form" onSubmit={handleSubmit}>
        <fieldset>
          <div className="pure-control-group">
            <label htmlFor="optOne">Option 1: </label>
            <input
              className="pure-input-2-3"
              type="text"
              id="optOne"
              value={options[0]}
              onChange={handleChange(0)}
              placeholder="Type your first option here..."
            />
          </div>
          <div className="pure-control-group x-margin">
            <label htmlFor="optTwo">Option 2: </label>
            <input
              className="pure-input-2-3"
              type="text"
              id="optTwo"
              value={options[1]}
              onChange={handleChange(1)}
              placeholder="Type your second option here..."
            />
          </div>
          <div className="pure-controls">
            <button
              className="pure-button pure-button-primary"
              disabled={options.includes("")}
              type="submit"
            >
              Submit
            </button>
          </div>
        </fieldset>
      </form>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    id: state.authedUser,
  };
};

export default connect(mapStateToProps)(CreateQuestion);
