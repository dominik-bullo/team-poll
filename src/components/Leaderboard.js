import { connect } from "react-redux";
import { Box } from "@mui/material";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TableRow from "@mui/material/TableRow";
import Paper from "@mui/material/Paper";
import Avatar from "@mui/material/Avatar";
import { Typography } from "@mui/material";

const Leaderboard = (props) => {
  return (
    <Box>
      <Typography align="center" variant="h5" component="h2" mb={2}>
        Leaderboard
      </Typography>
      <TableContainer component={Paper}>
        <Table sx={{ minWidth: 650 }} aria-label="simple table">
          <TableHead>
            <TableRow>
              <TableCell>#</TableCell>
              <TableCell>Avatar</TableCell>
              <TableCell>User</TableCell>
              <TableCell>Answered</TableCell>
              <TableCell>Created</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {props.userObjects.map((user, index) => (
              <TableRow key={user.id}>
                <TableCell>{index + 1}</TableCell>
                <TableCell>
                  <Avatar src={user.avatarURL} alt={`${user.name}'s avatar`} />
                </TableCell>
                <TableCell>{user.name}</TableCell>
                <TableCell>{user.answered}</TableCell>
                <TableCell>{user.created}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </Box>
  );
};

const mapStateToProps = ({ users }) => {
  const usersIds = Object.keys(users);
  const userObjects = usersIds
    .map((userId) => ({
      id: userId,
      avatarURL: users[userId].avatarURL,
      name: users[userId].name,
      answered: Object.keys(users[userId].answers).length,
      created: users[userId].questions.length,
    }))
    .sort((a, b) => b.created + b.answered - (a.created + a.answered));
  return { userObjects };
};

export default connect(mapStateToProps)(Leaderboard);
