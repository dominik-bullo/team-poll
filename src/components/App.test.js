import React from "react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";

import store from "../store/store";
import CreateQuestion from "./CreateQuestion";
import { _saveQuestion, _saveQuestionAnswer } from "../utils/_DATA";
import App from "../components/App";
import Leaderboard from "../components/Leaderboard";

describe("_saveQuestion (function)", () => {
  it("the saved question is returned and all expected fields are populated when correctly formatted data is passed", async () => {
    const question = {
      optionOneText: "Option One",
      optionTwoText: "Option Two",
      author: "someUser",
    };
    const result = await _saveQuestion(question);
    expect(result).toEqual({
      id: expect.any(String),
      timestamp: expect.any(Number),
      author: question.author,
      optionOne: {
        votes: [],
        text: question.optionOneText,
      },
      optionTwo: {
        votes: [],
        text: question.optionTwoText,
      },
    });
  });
  it("an error is returned if incorrect data is passed", async () => {
    const question = {
      optionOneText: "Option One",
      optionTwoText: "Option Two",
    };
    const result = _saveQuestion(question);
    await expect(result).rejects.toEqual(
      "Please provide optionOneText, optionTwoText, and author"
    );
  });
});

describe("_saveQuestionAnswer (function)", () => {
  it("true is returned when correctly formatted data is passed", async () => {
    const authedUser = "sarahedo";
    const qid = "8xf0y6ziyjabvozdd253nd";
    const answer = "optionOne";
    const result = await _saveQuestionAnswer({
      authedUser,
      qid,
      answer,
    });
    expect(result).toEqual(true);
  });
  it("an error is returned if incorrect data is passed", async () => {
    const authedUser = "sarahedo";
    const answer = "optionOne";
    const result = _saveQuestionAnswer({
      authedUser,
      answer,
    });
    await expect(result).rejects.toEqual(
      "Please provide authedUser, qid, and answer"
    );
  });
});

describe("App (component)", () => {
  it("displays user selection at login", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const input = view.getByTestId("user-select");
    expect(input).toBeInTheDocument();
  });
  it("displays button at login", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <App />
        </BrowserRouter>
      </Provider>
    );
    const button = view.getByTestId("login-button");
    expect(button).toBeInTheDocument();
  });
});

describe("CreateQuestion (component)", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <CreateQuestion />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("changes to input field are reflected in state", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
          <CreateQuestion />
        </BrowserRouter>
      </Provider>
    );
    const optionOneInput = getByTestId("option-one");
    const optionTwoInput = getByTestId("option-two");
    fireEvent.change(optionOneInput, { target: { value: "Option One" } });
    fireEvent.change(optionTwoInput, { target: { value: "Option Two" } });
    expect(optionOneInput.value).toBe("Option One");
    expect(optionTwoInput.value).toBe("Option Two");
  });

  it("allows submission only if all fields are filled", async () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <CreateQuestion />
        </BrowserRouter>
      </Provider>
    );
    const submitButton = view.getByTestId("submit-button");
    expect(submitButton).toBeDisabled();
    const optionOneInput = view.getByTestId("option-one");
    const optionTwoInput = view.getByTestId("option-two");
    fireEvent.change(optionOneInput, { target: { value: "Option One" } });
    fireEvent.change(optionTwoInput, { target: { value: "Option Two" } });
    expect(submitButton).not.toBeDisabled();
  });
});

describe("Leaderboard (component)", () => {
  it("shows table", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Leaderboard />
        </BrowserRouter>
      </Provider>
    );
    const table = view.getByTestId("leaderboard-table");
    expect(table).toBeInTheDocument();
  });
});
