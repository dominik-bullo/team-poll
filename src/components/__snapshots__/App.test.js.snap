// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateQuestion (component) matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="pure-u-1 new-question center"
  >
    <h2>
      New Question
    </h2>
    <h3>
      Would you rather...
    </h3>
    <form
      class="pure-form"
    >
      <fieldset>
        <div
          class="pure-control-group"
        >
          <label
            for="optOne"
          >
            Option 1: 
          </label>
          <input
            class="pure-input-2-3"
            data-testid="option-one"
            id="optOne"
            placeholder="Type your first option here..."
            type="text"
            value=""
          />
        </div>
        <div
          class="pure-control-group x-margin"
        >
          <label
            for="optTwo"
          >
            Option 2: 
          </label>
          <input
            class="pure-input-2-3"
            data-testid="option-two"
            id="optTwo"
            placeholder="Type your second option here..."
            type="text"
            value=""
          />
        </div>
        <div
          class="pure-controls"
        >
          <button
            class="pure-button pure-button-primary"
            data-testid="submit-button"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </div>
      </fieldset>
    </form>
  </div>
</DocumentFragment>
`;
